### PR TITLE
Allow any expression as join criteria for INNER joins

### PR DESCRIPTION
**Pull request by losipiuk**
Support multiple negations in conjuncts in JOIN condition clause

### CODE REVIEW

**Comment by losipiuk** [comment]
@kokosing please take a look

**Comment by losipiuk** [comment]
This is corresponding issue https://github.com/facebook/presto/issues/3888

**Comment by kokosing** [comment]
looks good

**Comment by losipiuk** [comment]
Addressed commits and rebased.

**Comment by losipiuk** [comment]
@cberner, can you take a look as you introduced initial non-equi join support?

**Comment by cberner** [comment]
I think this is too specific to `NOT` and `LIKE`. Instead lets solve the general case, so that it works for all expressions. I'll comment on the parts I think need to change, although there are probably some details to work out. Let me know if you have questions

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -1102,27 +1105,28 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                if (!isSupportedJoinConjunct(conjunct)) {
+                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported expression in join clause: %s", conjunct);
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I think the right way to do all of this would be to analyze each conjunct in context of the `output` `RelationType`. Then I think you'll be able to remove all this code with dependency extraction, and split the expressions into left and right

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 2]
```diff
@@ -227,31 +328,34 @@ protected RelationPlan visitJoin(Join node, Void context)
...
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                ComparisonExpression.Type comparisonType = comparison.getType();
-                if (comparison.getType() != EQUAL && node.getType() != INNER) {
+                JoinComparisonType comparisonType = JoinComparisonType.fromJoinConjunct(conjunct);
+                if (comparisonType != JoinComparisonType.EQUAL && node.getType() != INNER) {
                     throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins only supported for inner join: %s", conjunct);
                 }
-                Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(comparison.getLeft(), analysis.getColumnReferences());
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Same comment as in the analyzer. I think we can simplify this and make everything supported if we just divide the conjuncts into two lists: equalities, and everything else. Then everything else will just go into `postInnerJoinComparisons`

**Comment by cberner** [comment]
Let me know when you've updated!

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: @cberner. Thanks for comments.
I could use some help here :)

I tried to refactor the code into direction suggested by you but it have problems making it work. Either I misunderstood you or have to little understanding in how semantic analyzer/relation planner work (or both :) ).

I made two approaches:

1. I skipped the `DependencyExtractor` part in `SemanticAnalyzer` and `RelationPlanner`. Just performed analysis of all join condition conjuncts in context of `output` relation. Semantic analysis was passing but I got errors in planner.

I got error when (in case of equality conjunct) expression for left side expression was being added to plan for left join side. The code blow failed in `rewriteBase.getSymbol` as `analysis.getFieldIndex(node)` returned wrong index. And wrong index was stored in analysis object as analysis of join condition was done in context of `output` relation which different fields than `left` relation.

This is the part of `TranslationMap.rewriteExpressionWithResolvedName` which ultimately fails:
```java
       private Expression rewriteExpressionWithResolvedName(Expression node)
       {
           Optional<Integer> fieldIndex = analysis.getFieldIndex(node);
           checkState(fieldIndex.isPresent(), "No field mapping for node '%s'", node);

           Symbol symbol = rewriteBase.getSymbol(fieldIndex.get());
```

Sandbox code for this approach is here https://github.com/losipiuk/presto/tree/non-equi-join-sandbox.

2. In second approach I changed `ExpressionAnalyzer` in such way that we could pass more than one `RelationType` as `tupleDescriptor` parameter to `analyze` method.
Then in places where `tupleDescriptor.indexOf` and `tupleDescriptor.resolveFields` was called, I iterated over passed `tupleDescriptors` and returned result from one which could provide it. There was assumption that field sets in provided list of tupleDiscriptors was disjoint so only one should return result.

With this change planning finished but I got errors from `PlanSanityChecker` 

For query:
```
SELECT COUNT(clerk) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey APPROXIMATE AT 99.999 CONFIDENCE"
```
I got following error:
```
java.lang.IllegalArgumentException: Invalid node. Predicate dependencies ([lineitem, orders]) not in source plan output ([orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, shipdate, commitdate, receiptdate, shipinstruct, shipmode, comment, row_number, $sampleweight, expr, orderkey_0, custkey, orderstatus, totalprice, orderdate, orderpriority, clerk, shippriority, comment_1, row_number_2, $sampleweight_3, expr_4])
```

Not yet sure why this is happening. 
Sandbox code for this approach is here: https://github.com/losipiuk/presto/tree/non-equi-join-sandbox2


I would be super helpful for any hints on how to tackle the problem. Do you still think that analysing join expression in context of output relation is right choice? It seem proper to me but I am not sure how vast changes such approach require. Nevertheless it will probably be needed anyway, if we want to support complex expressions, other that inequalities with one side of inequality only touching columns from one side of join, as join condition.


One more thing concerning your comment. Could you elaborate on this sentence:
```
Then I think you'll be able to remove all this code with dependency extraction, and split the expressions into left and right
```

How do you envision splitting into left and right subexpressions of join condition for such query
```
select * from a join b on f(a.x, b.y) = 1
``` 
? what would be the result of splitting?
Are there any tools in the codebase already to do such splitting?

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Approach 1 seems like the right way to go to me, and I think I know why it's not working for you. It's slightly more complicated than I realized.
First, a bit of background:
The planner uses `TranslationMap` to convert field references into `Symbol`. You can see this in `TranslationMap.translateNamesToSymbols`. For this to work correctly, it needs the `analysis` to contain the correct field index into the relation. This whole process could be greatly simplified, but that's just how it works at the moment.

The reason that stopped working is that you analyzed the EQUALs expressions using the `output` relation type in `StatementAnalyzer`, but then the `TranslationMap` that gets used in the planner is from the left/right relation. To make this work I think you'll need to do the following, in both `StatementAnalyzer` and `RelationPlanner`:
* for EQUALs comparisons, use the `DependencyExtractor` to split the expression and then analyze/translate it using the left/right relation as appropriate
* for all other expressions (since they're going to be added as a post-JOIN filter step), analyze/translate them using the output relation from the JOIN. To do this in the planner it looks like you'll need to create a new `PlanBuilder` after this line `PlanNode root = new JoinNode(idAllocator.getNextId()` so that you can use that to rewrite the post-JOIN filter expressions. I think that will just work, but there may be some more issue there that we have to work through.

Let me know if you run into more issues!

**Comment by losipiuk** [comment]
Update. 

In this patch I treat all the expressions in form of `e1 OP e2`, where symbols from left and right join side are not mixed in neither e1 nor e2, old way. E1 and e2 computed as part of output for LEFT and RIGHT relation.

Conjuncts which do not match above schema are pushed down as post-join-condition after rewriting to use symbols from LEFT and RIGHT relation.

@cberner, please take a look and let me know what do you think.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 3]
```diff
@@ -316,9 +312,36 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
             root = new ProjectNode(idAllocator.getNextId(), root, projections.build());
         }
 
+        // rewrite all the other conditions using output symbols from left + right plan node.
+        RelationPlan intermediateRootRelationPlan = new RelationPlan(root, outputDescriptor, outputSymbols, sampleWeight);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: It would be nice to not build this intermediateRootRelationPlan which is only used to build TranslationMap below.
Is that possible?

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: @martint any thoughts on a cleaner way to do this?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
...
-                    rightExpression = comparison.getLeft();
-                }
-                else {
-                    // must have a complex expression that involves both tuples on one side of the comparison expression (e.g., coalesce(left.x, right.x) = 1)
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                    // analyze the clauses to record the types of all subexpressions and resolve names against the left/right underlying tuples
+                    if (rightExpression != null) {
+                        ExpressionAnalysis leftExpressionAnalysis = analyzeExpression(leftExpression, left, context);
+                        ExpressionAnalysis rightExpressionAnalysis = analyzeExpression(rightExpression, right, context);
+                        analysis.addJoinInPredicates(node, new Analysis.JoinInPredicates(leftExpressionAnalysis.getSubqueryInPredicates(), rightExpressionAnalysis.getSubqueryInPredicates()));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I discovered a bug in this code, while reviewing your PR (https://github.com/facebook/presto/issues/4250). Can you take a look at that? I'm concerned that your changes could introduce a correctness problem, where a query like the one in that issue returns an incorrect answer rather than failing.

File: presto-main/src/main/java/com/facebook/presto/sql/JoinExpressionUtils.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+package com.facebook.presto.sql;
+
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.LikePredicate;
+import com.facebook.presto.sql.tree.NotExpression;
+
+import java.util.Optional;
+
+public class JoinExpressionUtils
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/JoinExpressionUtils.java]: Utility classes should be `final`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I think this normalization can be removed (or moved to the `SimplifyExpressions` optimizer). It's only there because we didn't support non-`ComparisonExpression`s before

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
...
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
-                }
-
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(comparison.getLeft(), analyzer.getColumnReferences());
-                Set<QualifiedName> secondDependencies = DependencyExtractor.extractNames(comparison.getRight(), analyzer.getColumnReferences());
+                analyzeExpression(conjunct, output, context);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Why do we need this? Add a comment

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
...
+                if (isComparisionExpression(conjunct)) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Shouldn't there be an `else` clause to handle the case when it's not a comparison expression?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 9]
```diff
@@ -219,47 +219,49 @@ protected RelationPlan visitJoin(Join node, Void context)
...
-                if (firstDependencies.stream().allMatch(left.canResolvePredicate()) && secondDependencies.stream().allMatch(right.canResolvePredicate())) {
-                    leftExpression = comparison.getLeft();
-                    rightExpression = comparison.getRight();
-                }
-                else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && secondDependencies.stream().allMatch(left.canResolvePredicate())) {
-                    leftExpression = comparison.getRight();
-                    rightExpression = comparison.getLeft();
-                    comparisonType = flipComparison(comparisonType);
+
+                if (isComparisionExpression(conjunct)) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I think this code can be simplified a lot by only using this code path for `=` and treating all other `ComparisonExpression` as `postInnerJoinConditions`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 10]
```diff
@@ -219,47 +219,49 @@ protected RelationPlan visitJoin(Join node, Void context)
                 .build();
 
         ImmutableList.Builder<JoinNode.EquiJoinClause> equiClauses = ImmutableList.builder();
-        Expression postInnerJoinCriteria = new BooleanLiteral("TRUE");
+        List<Expression> nonComparisonExpressions = new ArrayList<>();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I think you can remove this and just have the `postInnerJoinConditions` list, if you change the analyzer to only have special treatment for `=`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This normalization is needed because we want to effectively handle expressions like 
`JOIN ON NOT a.x = b.x`. Without normalization the join condition would be put in postInnerJoin conditions which is not what we want.

But you are are probably right that this normalization should happen in different place - as part of standard rewrite flow. Currently I cannot put it in `SimplifyExpressions` as JoinNode does not contain original join condition; only list of `EquiJoinClause`s. And we do not have access to Analysis object in `SimplifyExpressions` rewriter. And even if had it, we probably do not want change it there. Right?

Did your intention was to add original join condition to JoinNode? And then change `SimplifyExpression` to include normalization and also add handling of JoinNodes?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 9 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Not sure. We would get rid of `joinConditionComparisonTypes` - but rest of this code is still needed for equality.

I would rather do this simplification as followup work.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 8 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I restructured the code a bit to make control flow more explicit.

**Comment by losipiuk** [comment]
I addressed most of the comments. Yet I will take a look at #4250.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: `normalize` on `NOT a.x = b.x` will result in `a.x != b.x` which will still go into `postInnerJoin`, unless I'm mistaken (see line 288 in `RelationPlanner`. It looks like `ComparisonExpression.Type.EQUAL` is the only type of predicate which has efficient handling

**Comment by cberner** [comment]
:+1: 

replied to your question about `NOT a.x = b.x` handling. Let me know when it's ready for me to take another look

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Yeah - right. There still can be sth like `NOT(NOT(a.x = b.x))`. But maybe iwe can assume that it so rare that we do not want to care about it. 

And if we still wanted to keep normalization but move it to `SimplifyExpression` rewrite - does what I wrote make sense? Do we want to add original join condition into `JoinNode`? Or you thought about another approach?

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: And one more comment on keeping the specific code for any comparison operator (not only equality). 

In case when expression corresponding to right (replicated) table is expensive to compute (CPU-wise) - with current approach we may be more efficient. Now computation is done once, on machine which executes right relation. If we always used postInnerJoin it would be computed multiple times for same rows - on machines performing JOIN.

Just my 2 cents. Probably not very practical scenario.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I think `NOT(NOT(a.x = b.x))` is pretty unlikely. That said I *think* it will just work if you move it to `SimplifyExpressions`. What should happen is that `SimplifyExpressions` will change it to `a.x = b.x` and then the `PredicatePushDown` optimizer should move it into the `JoinNode` as an `EquiJoinClause`. Be sure to test though, as I'm less familiar with the push down optimizer.

In general, I think the planners should be designed to generate the simplest possible plan which is correct, so I'd even go so far as to say that the planner shouldn't be putting anything in the `EquiJoinClause` and should plan it as a `CROSS JOIN` followed by a `Filter`, and then leave it to the `PredicatePushDown` optimizer to "fix" it.

This has the advantage that even if I write my query as:
```
SELECT *
FROM a
CROSS JOIN b
WHERE a.x = b.x
```
The optimizer could convert it to an `INNER JOIN` (I don't think we do this yet)

**Comment by losipiuk** [comment]
I added commit fixing #4250

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: What you write makes sense. Though if, I understand correctly changing the code the way you suggest, is total rebuild of how join is planned today. We would need to add original join condition into `JoinNode` (it is not there right now - only list of `EquiJoinClause`s). And then move most of the logic from `RelationPlanner.visitJoin` to one or more optimizers.

As currently all the detailed planning happens in `RelationPlanner.visitJoin` I cannot move the normalization into `SimplifyExpressions` as this one is executing *after* `RelationPlanner.visitJoin`. 

I would rather not put this Join planning refactor as part of this PR. Honestly I am not sure if with my current knowledge I would be able to efficiently implement this at all :)

**Comment by losipiuk** [comment]
This code requires rebase. We no longer have `Analysis.addInPredicates` method. I will complete this on Monday or over the weekend.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Ya, lets not make major changes to the planning of JOIN in this PR. It
shouldn't require adding the original predicate to the JoinNode when we do
though. It can be planned as a join followed by a filter, and rely on
predicate push down.

I think the normalization of not not a.x = b.x can be done pretty easily in
the optimizers, in the way I mentioned earlier

**Comment by cberner** [comment]
Not sure what happened to my comment, so posting it again:
> What you write makes sense. Though if, I understand correctly changing the code the way you suggest, is total rebuild of how join is planned today. We would need to add original join condition into JoinNode (it is not there right now - only list of EquiJoinClauses). And then move most of the logic from RelationPlanner.visitJoin to one or more optimizers.
>
> As currently all the detailed planning happens in RelationPlanner.visitJoin I cannot move the normalization into SimplifyExpressions as this one is executing after RelationPlanner.visitJoin.
>
> I would rather not put this Join planning refactor as part of this PR. Honestly I am not sure if with my current knowledge I would be able to efficiently implement this at all :)

Ya, lets not make major changes to the planning of JOIN in this PR. It shouldn't require adding the original predicate to the JoinNode when we do though. It can be planned as a join followed by a filter, and rely on predicate push down. 

I think the normalization of not not a.x = b.x can be done pretty easily in the optimizers, in the way I mentioned earlier

**Comment by losipiuk** [comment]
@cberner. I pushed another changeset after rebase. 

The possible additional which would not bloat the scope are:
 * remove `normalize`. I can do this. But I do not feel like it is correct approach unless we reintroduce it in optimizers. Which not super trivial. Keeping `normalize` does not help much - but helps a bit. And does not pollute code very much.
 * Only compute expression on joined relation sides for `=` operator and put all other operator in `postJoinConditions`. But I would not say that it will make code much simpler. We will not need the list of operator types. Rest would be still needed. This change is trivial. I can do this if you insist - but here again I do not see a big benefit from this change.

Please take a look when you have time and let me know if you want some more changes to this PR.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 11]
```diff
@@ -1102,38 +1103,52 @@ else if (criteria instanceof JoinOn) {
...
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                    // non-comparison expression.
+                    // expression will be put in post-join condition; analyze in context of output table.
+                    postJoinConjuncts.add(conjunct);
                 }
-
-                // analyze the clauses to record the types of all subexpressions and resolve names against the left/right underlying tuples
-                ExpressionAnalysis leftExpressionAnalysis = analyzeExpression(leftExpression, left, context);
-                ExpressionAnalysis rightExpressionAnalysis = analyzeExpression(rightExpression, right, context);
-                addCoercionForJoinCriteria(node, leftExpression, rightExpression);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Do coercions still work without this?

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: It seems so. Basted on tests :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 12]
```diff
@@ -271,23 +272,21 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
...
-            for (int i = 0; i < comparisonTypes.size(); i++) {
-                Symbol leftSymbol = leftPlanBuilder.translate(leftExpressions.get(i));
-                Symbol rightSymbol = rightPlanBuilder.translate(rightExpressions.get(i));
+            for (int i = 0; i < leftComparisonExpressions.size(); i++) {
+                Symbol leftSymbol = leftPlanBuilder.translate(leftComparisonExpressions.get(i));
+                Symbol rightSymbol = rightPlanBuilder.translate(rightComparisonExpressions.get(i));
 
-                if (comparisonTypes.get(i) == ComparisonExpression.Type.EQUAL) {
+                if (joinConditionComparisonTypes.get(i) == ComparisonExpression.Type.EQUAL) {
                     equiClauses.add(new JoinNode.EquiJoinClause(leftSymbol, rightSymbol));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Seems like there should be a `continue` here. (You can put that in a separate commit though, since it's a pre-existing bug)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 13]
```diff
@@ -219,47 +218,49 @@ protected RelationPlan visitJoin(Join node, Void context)
...
+                        joinConditionComparisonTypes.add(comparisonType);
+                    }
+                    else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && secondDependencies.stream().allMatch(left.canResolvePredicate())) {
+                        leftComparisonExpressions.add(secondExpression);
+                        rightComparisonExpressions.add(firstExpression);
+                        joinConditionComparisonTypes.add(flipComparison(comparisonType));
+                    }
+                    else {
+                        // the case when we mix symbols from both left and right join side on either side of condition.
+                        nonComparisonExpressions.add(conjunct);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: This name doesn't seem right. The variable is named `nonComparisonExpressions` but the only expressions you add to it are ones that pass `conjunct instance of ComparisonExpression`. Maybe name it `complexComparisonExpressions`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 14]
```diff
@@ -1102,38 +1103,52 @@ else if (criteria instanceof JoinOn) {
             analyzer.analyze((Expression) optimizedExpression, output, context);
             analysis.addCoercions(analyzer.getExpressionCoercions());
 
+            Set<Expression> postJoinConjuncts = new HashSet<>();
+            final Set<InPredicate> leftJoinInPredicates = new HashSet<>();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: `final` not needed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 11 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Ah, I debugged this to see what was happening and it's because we call `analysis.addCoercions(analyzer.getExpressionCoercions());`  after reanalyzing the optimized expression

**Comment by cberner** [comment]
Couple minor comments

**Comment by cberner** [comment]
@martint can you look at the one part I tagged you in?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 13 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I named it `complexJoinExpressions`. Not super nice. But I do not like `complexComparisonExpressions` neither as it stores not only `ComparisonExpressions`.

**Comment by losipiuk** [comment]
Adressed comments and rebased onto master.

File: presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java [hunk 15]
```diff
@@ -258,6 +258,9 @@ public static Expression normalize(Expression expression)
                 ComparisonExpression comparison = (ComparisonExpression) not.getValue();
                 return new ComparisonExpression(negate(comparison.getType()), comparison.getLeft(), comparison.getRight());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java]: this returns new expression which might not be matched in IdentityHashMap in Analysis (see getType() method).

**Comment by martint** [comment]
Can you rebase this on master?

**Comment by losipiuk** [comment]
Rebased - travis build in progress

**Comment by martint** [comment]
Merged, thanks!
